src/plots/linePlot.ts

```diff
@@ -425,6 +449,72 @@ export module Plots {
       return filteredDataIndices;
     }
 
+    private _filterDownsampling(dataset: Dataset, indices: number[]) {
+      let data = dataset.data();
+      let filteredIndices: number[] = [];
+      let minScaledValue: number;
+      let maxScaledValue: number;
+      let currentSlope: number;
+      let scaledXAccessor = Plot._scaledAccessor(this.x());
+      let scaledYAccessor = Plot._scaledAccessor(this.y());
+
+      if (indices.length === 0) {
+        return filteredIndices;
+      }
+      filteredIndices.push(indices[0]);
+
+      let belongingToCurrentBucket = (i: number) => {
```

Comment:
I feel the function itself defines bucket. In the function, we put the end points of adjacent lines that have same slopes into one bucket.